python/pyiceberg/table/__init__.py

```diff
@@ -69,21 +72,288 @@
     import ray
     from duckdb import DuckDBPyConnection
 
+    from pyiceberg.catalog import Catalog
 
 ALWAYS_TRUE = AlwaysTrue()
 
 
+class TableUpdates:
```

Comment:
I think the rest of this PR is about ready, my remaining concern is this API. I don't think that we want to expose methods that directly set schema, partitioning, or other table metadata. Schema is a great example because people are going to really mess up the field IDs. We don't want to encourage that through the API.

For example, this is going to break:

```python
catalog = load_catalog(...)

schema1 = Schema(
    NestedField(field_id=0, name='id', field_type=LongType()),
    schema_id=1
)
schema2 = Schema(
    NestedField(field_id=0, name='id', field_type=LongType()),
    NestedField(field_id=1, name='data', field_type=StringType()),
    schema_id=1
)

table = catalog.create_table(identifier='test', schema=schema1)
table.set_schema(schema2).commit()
```

The table schema is incompatible with the second schema because the field IDs are reassigned.

Reply:
Good one. First of all, if we as a community decide that we really want to make things strict (which is a good thing), we should do checks on the server side that forbid this kind of operation. We can also add this on the Python side, for example, set the IDs of the existing fields (using the WithPartnerVisitor). People always will find ways to do weird stuff :D (for example, by copying Python snippets).

I would expect something like this:

```python
catalog = load_catalog(...)

schema = Schema(
    NestedField(field_id=0, name='id', field_type=LongType()),
    schema_id=1
)

table = catalog.create_table(identifier='test', schema=schema)

new_schema = table.schema.alter.add_field(
    name="data",
    type=StringType(),
    required=False
)

new_schema = table.schema.alter.drop_field(
    name="id"
)

table.set_schema(
    new_schema
)
```

However, this gets awkward for lists, maps, and structs.

Reply:
Would it be possible to add more tests later once we get the schema, partitioning, and sorting figured out? I did some refactoring on the integration tests. But that's being blocked on this one: https://github.com/apache/iceberg/pull/7704

Reply:
> Would it be possible to add more tests later once we get the schema, partitioning, and sorting figured out?

Yes! I think we should limit this to setting and removing properties. Then we can handle the schema, spec, and write order changes later.

> However, this gets awkward for lists, maps, and structs.

We already handle these cases in the Java API, using the same approach that you outlined above with methods to make individual changes to the schema that correspond to SQL DDL operations (e.g. ADD COLUMN). There are [thorough tests](https://github.com/apache/iceberg/blob/master/core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java#L669-L737) as well.

Reply:
Nice, I can add that. Thanks for pointing out the Java tests. 

I had one follow-up PR on the integration tests: https://github.com/apache/iceberg/pull/7768 Would be nice to get those in a well.